idea_plugin/src/com/google/googlejavaformat/intellij/FormatterUtil.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.googlejavaformat.intellij;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.collect.BoundType;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Range;
+import com.google.googlejavaformat.java.Formatter;
+import com.google.googlejavaformat.java.FormatterException;
+import com.intellij.openapi.util.TextRange;
+import java.util.Collection;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+final class FormatterUtil {
+
+  private FormatterUtil() {}
+
+  static Map<TextRange, String> getReplacements(
+      Formatter formatter, String text, Collection<TextRange> ranges) {
+    try {
+      ImmutableMap.Builder<TextRange, String> replacements = ImmutableMap.builder();
+      formatter
+          .getFormatReplacements(text, toRanges(ranges))
+          .forEach(
+              replacement ->
+                  replacements.put(
+                      toTextRange(replacement.getReplaceRange()),
+                      replacement.getReplacementString()));
+      return replacements.build();
```

Comment:
@plumpy I believe that the code in this try statement could be made more idiomatic with something like:

```java
return formatter
    .getFormatReplacements(text, toRanges(ranges))
    .stream()
    .collect(
        toImmutableMap(
            replacement -> toTextRange(replacement.getReplaceRange()),
            replacement -> replacement.getReplacementString()));
```

Reply:
This code is designed to run in IDEA 2017.1+, but 2017.1 has Guava 19, which does not have any Java 8 features like toImmutableMap.

Reply:
(and I'd happily bump to 2017.2 as a minimum version, but the current version of Android Studio is based on 2017.1, so I'd like to keep supporting that)

Reply:
Ah I see, that's a bit of a shame. Thanks for the clarification. :)